python/lsst/ip/diffim/zogy.py

```diff
@@ -0,0 +1,839 @@
+
+        This method performs ZOGY proper image subtraction on
+        `subExposure` using local measures for image variances and
+        PSF. `subExposure` is a sub-exposure of the science image. It
+        also requires the corresponding sub-exposures of the template
+        (`template`). The operations are actually performed on
+        `expandedSubExposure` to allow for invalid edge pixels arising
+        from convolutions.
+
+        Parameters
+        ----------
+        subExposure : lsst.afw.image.Exposure
+            the sub-exposure of the diffim
+        expandedSubExposure : lsst.afw.image.Exposure
+            the expanded sub-exposure upon which to operate
+        fullBBox : lsst.afw.geom.BoundingBox
+            the bounding box of the original exposure
+        template : lsst.afw.image.Exposure
+            the template exposure, from which a corresponding sub-exposure
+            is extracted
+        kwargs :
+            additional keyword arguments propagated from
+            `ImageMapReduceTask.run`. These include:
+        - doScorr : bool
+              Compute and return the corrected likelihood image S_corr
+              rather than the proper image difference
+        - inImageSpace : bool
+              Perform all convolutions in real (image) space rather than
+              in Fourier space. This option currently leads to artifacts
+              when using real (measured and noisy) PSFs, thus it is set
+              to `False` by default.
+            These kwargs may also include arguments to be propagated to
+            `ZogyTask.computeDiffim` and `ZogyTask.computeScorr`.
+
+        Returns
+        -------
+        A `lsst.pipe.base.Struct` containing the result of the
+        `subExposure` processing, labelled 'subExposure'. In this case
+        it is either the subExposure of the proper image difference D,
+        or (if `doScorr==True`) the corrected likelihood exposure `S`.
+
+        Notes
+        -----
+        This `run` method accepts parameters identical to those of
+        `ImageMapperSubtask.run`, since it is called from the
+        `ImageMapperTask`. See that class for more information.
+        """
+        bbox = subExposure.getBBox()
+        center = ((bbox.getBeginX() + bbox.getEndX()) // 2., (bbox.getBeginY() + bbox.getEndY()) // 2.)
+        center = afwGeom.Point2D(center[0], center[1])
+
+        imageSpace = kwargs.pop('inImageSpace', False)
+        doScorr = kwargs.pop('doScorr', False)
+        sigmas = kwargs.pop('sigmas', None)
+        padSize = kwargs.pop('padSize', 7)
+
+        # Psf and image for science img (index 2)
+        subExp2 = expandedSubExposure
+
+        # Psf and image for template img (index 1)
+        subExp1 = template.Factory(template, expandedSubExposure.getBBox())
+
+        if sigmas is None:
+            sig1 = sig2 = None
+        else:  # for testing, can use the input sigma (e.g., global value for entire exposure)
```

Comment:
Please put this comment and the following ones after `if` or `else` statements on their own line.